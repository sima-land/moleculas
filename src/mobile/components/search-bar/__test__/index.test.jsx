import React, { createRef } from 'react';
import ReactDOM from 'react-dom';
import { createEvent, fireEvent, render } from '@testing-library/react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { SearchBar } from '../index';
import ArrowLeftSVG from '@sima-land/ui-quarks/icons/24x24/Stroked/ArrowLeft';
import { DropdownItem } from '@sima-land/ui-nucleons/dropdown-item';

describe('SearchBar', () => {
  it('should renders properly with minimal props', () => {
    const onChange = jest.fn();
    const component = mount(<SearchBar onChange={onChange} />);
    expect(component).toMatchSnapshot();
    expect(component.find('.search-icon')).toHaveLength(1);
  });

  it('should renders properly with all props', () => {
    const onChange = jest.fn();
    const onBackClick = jest.fn();
    const onSearchClick = jest.fn();
    const onClearClick = jest.fn();
    const onCancelClick = jest.fn();
    const component = mount(
      <SearchBar
        value='Очень длинный запрос в поисковой строке'
        onChange={onChange}
        startButtons={[{ text: 'Назад', icon: ArrowLeftSVG, onClick: onBackClick }]}
        endButtons={[
          { text: 'Поиск', onClick: onSearchClick },
          { text: 'Отмена', onClick: onCancelClick },
        ]}
        withSearchIcon={false}
        inputMode='tel'
        onClear={onClearClick}
        description='150 324 предложений'
        placeholder='Найти'
      />,
    );
    expect(component).toMatchSnapshot();

    expect(component.find('.search-icon')).toHaveLength(0);
    expect(component.find('.search-field').prop('inputMode')).toBe('tel');
    expect(component.find('.search-field').prop('placeholder')).toBe('Найти');

    act(() => {
      component.find('.search-field').prop('onChange')({ target: { value: '123' } });
    });
    expect(onChange).toHaveBeenCalledTimes(1);

    act(() => {
      component.find('.clear-button').prop('onClick')();
      component.find('.before-field .button').prop('onClick')();
      component.find('.after-field .button').at(0).prop('onClick')();
      component.find('.after-field .button').at(1).prop('onClick')();
    });
    expect(onClearClick).toHaveBeenCalledTimes(1);
    expect(onBackClick).toHaveBeenCalledTimes(1);
    expect(onSearchClick).toHaveBeenCalledTimes(1);
    expect(onCancelClick).toHaveBeenCalledTimes(1);
  });

  it('should open and close dropdown on buttons handlers', () => {
    const onChange = jest.fn();
    const onCancelClick = jest.fn();
    const onSearchClick = jest.fn();
    const testEvent = jest.fn();
    const component = mount(
      <SearchBar
        value='Очень длинный запрос в поисковой строке'
        onChange={onChange}
        endButtons={[
          { text: 'Отмена', onClick: onCancelClick },
          { text: 'Поиск', onClick: onSearchClick },
        ]}
      />,
    );
    expect(component.find('.dropdown-container')).toHaveLength(0);

    act(() => {
      component.find('.narrow-screen .button').prop('onClick')();
    });
    component.update();

    expect(component.find('.dropdown-container')).toHaveLength(1);

    act(() => {
      component.find(DropdownItem).at(0).prop('onClick')(testEvent);
    });
    component.update();

    expect(onCancelClick).toHaveBeenCalledWith(testEvent);
    expect(component.find('.dropdown-container')).toHaveLength(0);
  });

  it('should close dropdown when resize window', () => {
    const onChange = jest.fn();
    const onCancelClick = jest.fn();
    const onSearchClick = jest.fn();
    const component = mount(
      <SearchBar
        value='Очень длинный запрос в поисковой строке'
        onChange={onChange}
        endButtons={[
          { text: 'Отмена', onClick: onCancelClick },
          { text: 'Поиск', onClick: onSearchClick },
        ]}
      />,
    );
    expect(component.find('.dropdown-container')).toHaveLength(0);

    act(() => {
      component.find('.narrow-screen .button').prop('onClick')();
    });
    component.update();
    expect(component.find('.dropdown-container')).toHaveLength(1);

    act(() => {
      window.dispatchEvent(new Event('resize'));
    });
    component.update();
    expect(component.find('.dropdown-container')).toHaveLength(0);
  });

  it('should close dropdown when outside click', () => {
    const onChange = jest.fn();
    const onCancelClick = jest.fn();
    const onSearchClick = jest.fn();

    const container = document.createElement('div');
    document.body.appendChild(container);

    act(() => {
      ReactDOM.render(
        <SearchBar
          value='Очень длинный запрос в поисковой строке'
          onChange={onChange}
          endButtons={[
            { text: 'Отмена', onClick: onCancelClick },
            { text: 'Поиск', onClick: onSearchClick },
          ]}
        />,
        container,
      );
    });
    expect(container.querySelectorAll('.dropdown-container')).toHaveLength(0);
    act(() => {
      container.querySelector('.narrow-screen .button').click();
    });
    expect(container.querySelectorAll('.dropdown-container')).toHaveLength(1);
    act(() => {
      container.querySelector('.search-field').click();
    });
    expect(container.querySelectorAll('.dropdown-container')).toHaveLength(0);

    ReactDOM.unmountComponentAtNode(container);
  });

  it('should handle "inputRef" prop', () => {
    const ref = createRef();

    mount(<SearchBar inputRef={ref} value='Запрос в поисковой строке' onChange={jest.fn()} />);

    expect(ref.current instanceof HTMLInputElement).toBe(true);
  });

  it('should prevent input blur by prevent input wrapper mouse down', () => {
    const { getByTestId } = render(
      <SearchBar value='Hello!' onChange={jest.fn()} onClear={jest.fn()} />,
    );

    fireEvent.focus(getByTestId('search-bar-input'));

    const event = createEvent.mouseDown(getByTestId('search-bar:input-wrapper'));

    expect(event.defaultPrevented).toBe(false);
    fireEvent(getByTestId('search-bar:input-wrapper'), event);
    expect(event.defaultPrevented).toBe(true);
  });

  it('should NOT prevent input blur when its', () => {
    const { getByTestId } = render(
      <SearchBar value='Hello!' onChange={jest.fn()} onClear={jest.fn()} />,
    );

    const event = createEvent.mouseDown(getByTestId('search-bar:input-wrapper'));
    Object.defineProperty(event, 'target', {
      value: getByTestId('search-bar-input'),
    });
    fireEvent(getByTestId('search-bar:input-wrapper'), event);

    expect(event.defaultPrevented).toBe(false);
  });
});
