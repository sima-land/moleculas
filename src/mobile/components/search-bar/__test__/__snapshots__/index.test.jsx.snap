// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should renders properly with all props 1`] = `
<SearchBar
  description="150 324 предложений"
  endButtons={
    Array [
      Object {
        "onClick": [MockFunction],
        "text": "Поиск",
      },
      Object {
        "onClick": [MockFunction],
        "text": "Отмена",
      },
    ]
  }
  inputMode="tel"
  onChange={[MockFunction]}
  onClear={[MockFunction]}
  placeholder="Найти"
  startButtons={
    Array [
      Object {
        "icon": [Function],
        "onClick": [MockFunction],
        "text": "Назад",
      },
    ]
  }
  value="Очень длинный запрос в поисковой строке"
  withSearchIcon={false}
>
  <div
    className="wrapper"
  >
    <div
      className="before-field"
    >
      <div
        className="buttons-container"
      >
        <ForwardRef(Button)
          icon={[Function]}
          key="0"
          onClick={[MockFunction]}
          text="Назад"
        >
          <div
            aria-label="Назад"
            className="button"
            data-testid="search-bar-button"
            onClick={[MockFunction]}
            role="button"
            tabIndex={0}
          >
            <ArrowLeftSVG
              aria-hidden="true"
              fill="#212121"
            >
              <svg
                aria-hidden="true"
                fill="#212121"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M5.414 13l5.293 5.293a1 1 0 01-1.414 1.414l-7-7a1 1 0 010-1.414l7-7a1 1 0 011.414 1.414L5.414 11H21a1 1 0 110 2H5.414z"
                />
              </svg>
            </ArrowLeftSVG>
          </div>
        </ForwardRef(Button)>
      </div>
    </div>
    <label
      className="input-wrapper"
    >
      <div
        className="input-container"
      >
        <input
          autoFocus={true}
          className="search-field"
          data-testid="search-bar-input"
          inputMode="tel"
          onChange={[MockFunction]}
          placeholder="Найти"
          tabIndex={0}
          type="text"
          value="Очень длинный запрос в поисковой строке"
        />
        <div
          className="description-container"
        >
          <span
            className="description"
          >
            150 324 предложений
          </span>
        </div>
      </div>
      <ForwardRef(Link)
        aria-label="Очистить"
        className="clear-icon"
        data-testid="search-bar:clear-icon"
        onClick={[Function]}
        pseudo={true}
      >
        <a
          aria-label="Очистить"
          className="link clear-icon color__brand-blue"
          data-testid="search-bar:clear-icon"
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          <CrossSVG
            aria-hidden={true}
            fill="#9e9e9e"
          >
            <svg
              aria-hidden={true}
              fill="#9e9e9e"
              height={24}
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                clipRule="evenodd"
                d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zm4.95-6.464a1 1 0 11-1.414 1.414L12 13.414 8.464 16.95a1 1 0 11-1.414-1.414L10.586 12 7.05 8.464A1 1 0 018.464 7.05L12 10.586l3.536-3.536a1 1 0 111.414 1.414L13.414 12l3.536 3.536z"
                fillRule="evenodd"
              />
            </svg>
          </CrossSVG>
        </a>
      </ForwardRef(Link)>
    </label>
    <div
      className="after-field wide-screen"
    >
      <div
        className="buttons-container"
      >
        <ForwardRef(Button)
          key="0"
          onClick={[MockFunction]}
          text="Поиск"
        >
          <div
            className="button"
            data-testid="search-bar-button"
            onClick={[MockFunction]}
            role="button"
            tabIndex={0}
          >
            Поиск
          </div>
        </ForwardRef(Button)>
        <ForwardRef(Button)
          key="1"
          onClick={[MockFunction]}
          text="Отмена"
        >
          <div
            className="button"
            data-testid="search-bar-button"
            onClick={[MockFunction]}
            role="button"
            tabIndex={0}
          >
            Отмена
          </div>
        </ForwardRef(Button)>
      </div>
    </div>
    <div
      className="after-field narrow-screen"
    >
      <div
        className="buttons-container"
      >
        <ForwardRef(Button)
          icon={[Function]}
          onClick={[Function]}
          text="Больше"
        >
          <div
            aria-label="Больше"
            className="button"
            data-testid="search-bar-button"
            onClick={[Function]}
            role="button"
            tabIndex={0}
          >
            <MoreVertSVG
              aria-hidden="true"
              fill="#212121"
            >
              <svg
                aria-hidden="true"
                fill="#212121"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M10 5a2 2 0 114 0 2 2 0 01-4 0zM10 12a2 2 0 114 0 2 2 0 01-4 0zM12 17a2 2 0 100 4 2 2 0 000-4z"
                />
              </svg>
            </MoreVertSVG>
          </div>
        </ForwardRef(Button)>
      </div>
    </div>
  </div>
</SearchBar>
`;

exports[`SearchBar should renders properly with minimal props 1`] = `
<SearchBar
  onChange={[MockFunction]}
>
  <div
    className="wrapper"
  >
    <label
      className="input-wrapper"
    >
      <span
        className="search-icon"
      >
        <MagnifierSVG
          aria-hidden={true}
          fill="#9e9e9e"
        >
          <svg
            aria-hidden={true}
            fill="#9e9e9e"
            height={24}
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              clipRule="evenodd"
              d="M11 3a8 8 0 104.32 14.734l3.973 3.973a1 1 0 001.414-1.414l-3.847-3.847A8 8 0 0011 3zm-6 8a6 6 0 1112 0 6 6 0 01-12 0z"
              fillRule="evenodd"
            />
          </svg>
        </MagnifierSVG>
      </span>
      <div
        className="input-container"
      >
        <input
          autoFocus={true}
          className="search-field"
          data-testid="search-bar-input"
          onChange={[MockFunction]}
          placeholder="Поиск"
          tabIndex={0}
          type="text"
        />
      </div>
    </label>
  </div>
</SearchBar>
`;
